File: Week_02/cpp/src/_129_SumRootToLeafNumbers.cpp
#include <iostream>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution
{
public:
    int sumNumbers(TreeNode *root)
    {
        if (!root)
            return 0;
        int ret = 0, path = 0;
        dfs(root, &ret, &path);
        return ret;
    }
    void dfs(TreeNode *root, int *ret, int *path)
    {
        *path = (*path) * 10 + root->val;
        if (root->left)
            dfs(root->left, ret, path);
        if (root->right)
            dfs(root->right, ret, path);
        if (!root->left && !root->right)
            *ret += *path;
        *path /= 10;
    }
};

File: Week_02/kotlin/src/_129_SumRootToLeafNumbers.kt
class _129_SumRootToLeafNumbers_ {
    /**
     * Example:
     * var ti = TreeNode(5)
     * var v = ti.`val`
     * Definition for a binary tree node.
     * class TreeNode(var `val`: Int) {
     *     var left: TreeNode? = null
     *     var right: TreeNode? = null
     * }
     */
    class Solution {
        fun sumNumbers(root: TreeNode?): Int {
            root?:return 0
            var path = 0
            var ret = 0
            fun dfs(node: TreeNode) {
                path = path * 10 + node.`val`
                node.left?.let { dfs(it) }
                node.right?.let { dfs(it) }
                if (node.left == null && node.right == null) {
                    ret += path
                }
                path /= 10
            }
            dfs(root)
            return ret
        }
    }
}